mule-shaders-module/src/main/java/org/mule/module/shaders/api/ReflectiveAccessShader.java

```diff
@@ -0,0 +1,632 @@
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.shade.ShadeRequest;
+import org.apache.maven.plugins.shade.Shader;
+import org.apache.maven.plugins.shade.filter.Filter;
+import org.apache.maven.plugins.shade.relocation.Relocator;
+import org.apache.maven.plugins.shade.resource.ManifestResourceTransformer;
+import org.apache.maven.plugins.shade.resource.ReproducibleResourceTransformer;
+import org.apache.maven.plugins.shade.resource.ResourceTransformer;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.logging.AbstractLogEnabled;
+import org.codehaus.plexus.util.IOUtil;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.commons.Remapper;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PushbackInputStream;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+
+/**
+ *  A shader capable of changing the specified classes for reflective access before relocating packages.
+ *  Custom version of {@link org.apache.maven.plugins.shade.DefaultShader}.
+ */
+@Component(role = Shader.class, hint = "ReflectiveAccessShader")
+public class ReflectiveAccessShader
+    extends AbstractLogEnabled
+    implements Shader {
+
+  private static final int BUFFER_SIZE = 32 * 1024;
+
+  public void shade(ShadeRequest shadeRequest)
+      throws IOException, MojoExecutionException {
+    Set<String> resources = new HashSet<>();
+
+    ManifestResourceTransformer manifestTransformer = null;
+    List<ResourceTransformer> transformers =
+        new ArrayList<>(shadeRequest.getResourceTransformers());
+    for (Iterator<ResourceTransformer> it = transformers.iterator(); it.hasNext();) {
+      ResourceTransformer transformer = it.next();
+      if (transformer instanceof ManifestResourceTransformer) {
+        manifestTransformer = (ManifestResourceTransformer) transformer;
+        it.remove();
+      }
+    }
+
+    RelocatorRemapper remapper = new RelocatorRemapper(shadeRequest.getRelocators());
+
+    // noinspection ResultOfMethodCallIgnored
+    shadeRequest.getUberJar().getParentFile().mkdirs();
+
+    JarOutputStream out = null;
+    try {
```

Comment:
try with resources.

Reply:
Since this file was copied from `org.apache.maven.plugins.shade.DefaultShader` and we have since confirmed with legal that we should open source the project under Apache v2.0, it makes more sense to keep this file as close to the original as possible, in order to make it easier to maintain for future versions of the plugin.